### ISSUE TITLE
Semantic data remains after changing page to redirect

### ISSUE THREAD

**Issue by darenwelsh** [SemanticMediaWiki/SemanticMediaWiki#895]
Possibly related to [#869] (https://github.com/SemanticMediaWiki/SemanticMediaWiki/issues/869) and [#566] (https://github.com/SemanticMediaWiki/SemanticMediaWiki/issues/566).

1. Create page using a template that sets semantic properties for that page
 * Browse properties for that page and note semantic data
2. Revise that page so it only has code to redirect to a different page
 * Browse properties for that page again and note that semantic data remains

See [this example] (http://semantic-mediawiki.org/w/index.php?title=User:Darenwelsh/Test_Page_2&redirect=no) and note how [its properties] (http://semantic-mediawiki.org/wiki/Special:Browse/User:Darenwelsh-2FTest_Page_2) remained after I revised it to just redirect to another page.

This also happens on pages that use Semantic Forms, but not on pages where a semantic property is set without use of a template. For example

1. Create page including some kind of property being set with {{#set: myProperty = value }}
 * Browse properties for that page and note semantic data
2. Revise that page so it only has code to redirect to a different page
 * Browse properties for that page again and note that semantic data is gone

See [this example] (http://semantic-mediawiki.org/w/index.php?title=User:Darenwelsh/Test_Page_1&redirect=no) and note how [its properties] (http://semantic-mediawiki.org/wiki/Special:Browse/User:Darenwelsh-2FTest_Page_1) are gone as they should be now that the page is just a redirect to another page.

**Comment by mwjames** [comment]
One could cry, it has to be the redirects that consistently create issues and this time it seems to be connected with template usage.

PS: Currently, I don't feel in the mood for a bug hunt (and its sounds like a real pain bug) the last time it cost me a week to create a proper use case let alone to track down the #347 issue.

**Comment by mwjames** [comment]
Here some hints for people who want to debug this, the move/redirect action is initiated by [0] which eventually calls [1] `SMWSQLStore3Writers ::changeTitle` to reshuffle the internal Id assignments.

[0] https://github.com/SemanticMediaWiki/SemanticMediaWiki/blob/master/src/MediaWiki/Hooks/TitleMoveComplete.php

[1] https://github.com/SemanticMediaWiki/SemanticMediaWiki/blob/master/includes/storage/SQLStore/SMW_SQLStore3_Writers.php

**Comment by mwjames** [comment]
The above statements need refinement in its description because using `&action=move` or using `#REDIRECT` (to edit manually) will produce different results.

## Use case refinement 
Template:UseTemplateForAnnotation
```
<includeonly>
* Has page: [[Has page::{{PAGENAME}}]]
* Has number: [[Has number::100134]]
* Has quantity: [[Has quantity::101.25 km²]]
* Has date: [[Has date::1 Jan 2114]]
* Has Url: [[Has Url::http://loremipsum.org/]]
* Has annotation uri: [[Has annotation uri::http://loremipsum.org/foaf.rdf]]
* Has email: [[Has email::Lorem@ipsum.org]]
* Has temperature: [[Has temperature::1021 °C]]
* Has wattage: [[Has wattage::90011]], [[Has wattage::9002 W]], [[Has wattage::10 kW]]
* Has telephone number: [[Has telephone number::+1-2012-555-0123]]
* Has record: [[Has record::Lorem ipsum; closed]]
* Has status: [[Has status::open]]
* Has boolean: [[Has boolean::true]]

{{#subobject:
 |Has page={{PAGENAME}}
 |Has number=1111
 |Has quantity=25 sqmi
 |Has date=January 4, 2010 7:21 pm
 |Has Url=http://example.org/some/
 |Has annotation uri=http://example.org/foaf.rdf
 |Has email=Lorem@example.org
 |Has temperature=1002 °F
 |Has wattage=42
 |Has wattage=42 hp
 |Has telephone number=+1-201-555-5555
 |Has record=Lorem enim; closed
 |Has status=closed
 |Has boolean=false
 |has page=Foo-A,Foo-CC|+sep=,
}}
</includeonly>
```

### Page with `&action=move` to redirect to unknown location

* Create page ABC/1 with `{{UseTemplateForAnnotation}}` content
* Move page ABC/1 to an unkown ABC/2
* This should move all annotations to the  ABC/2 leaving `ABC/2` as redirect

### Page with `&action=move` to redirect to known location

* Create page ABC/3 with `{{UseTemplateForAnnotation}}` content
* Move page ABC/3 to known ABC/4
* This should move all annotations to the ABC/4 leaving `ABC/3` as redirect

### Page with `#REDIRECT` edit to redirect to unknown location

* Create page DEF/1 with `{{UseTemplateForAnnotation}}` content
* Edit page DEF/1 to unkown `#REDIRECT [[DEF/2]]`
* This does not create DEF/2 and therefore all annotations remain at DEF/1

### Page with `#REDIRECT` edit to redirect to known location

* Create page DEF/3 with `{{UseTemplateForAnnotation}}` content
* Edit page DEF/3 to known `#REDIRECT [[DEF/4]]`
* This does not move content from DEF/3 to DEF/4 therefore all annotations remain at DEF/3

@darenwelsh Can you confirm above observation?

**Comment by darenwelsh** [comment]
I tried this on our dev EVA wiki. I didn't want to create a bunch of new test pages on SMW, but since my results were different, maybe we should.

I assume "unknown" means a page that does not exist and that "known" means the page already exists. Please correct me if I'm wrong.

### Page with `&action=move` to redirect to unknown location
* Created page ABC/1 with `{{UseTemplateForAnnotation}}` content
* Moved page ABC/1 to an unknown ABC/2 (left a redirect behind)
* This copied all annotations to ABC/2, left ABC/1 as redirect to ABC/2 with annotations also remaining on ABC/1

### Page with `&action=move` to redirect to known location
* Created page ABC/3 with `{{UseTemplateForAnnotation}}` content
* Moved page ABC/3 to known ABC/4 (deleted page ABC/4 to allow moving ABC/3 to ABC/4)
* This copied all annotations to ABC/4, left ABC/3 as redirect to ABC/4 with annotations also remaining on ABC/3

### Page with `#REDIRECT` edit to redirect to unknown location
* Created page DEF/1 with `{{UseTemplateForAnnotation}}` content
* Edited page DEF/1 to unkown `#REDIRECT [[DEF/2]]`
* This does not create DEF/2 and therefore all annotations remain at DEF/1 **(confirmed)**

### Page with `#REDIRECT` edit to redirect to known location
* Created page DEF/3 with `{{UseTemplateForAnnotation}}` content
* Edited page DEF/3 to known `#REDIRECT [[DEF/4]]`
* This does not move content from DEF/3 to DEF/4 therefore all annotations remain at DEF/3 **(confirmed)**

**Comment by mwjames** [comment]
@kghbln @darenwelsh Addressing both of you as I need some assistance to verify that the "hack" works and doesn't introduce any other regression.

## Issue

If a page is moved (or a redirect is created) via `Special:MovePage` then MW's internal hook is activated which calls `SMWSQLStore3Writers::changeTitle` (the function that moves the Id's).

Using `#REDIRECT` circumvents any functionality that involves `SMWSQLStore3Writers ::changeTitle` and this is the reason why we see data still appearing on the page that is supposed to be redirected and a difference in behaviour of  `#REDIRECT`/`Special:MovePage`.

As for the use of `#REDIRECT`, `SMWSQLStore3Writers::doFlatDataUpdate` has a check for redirects which includes the following comment:
```
		// Stop here:
		// * no support for annotations on redirect pages
		// * updateRedirects takes care of deleting any previous data
```
Unfortunately this comment is not correct (especially in those cases where the redirect target is unknown/doesn't exist) and this is why we see data remaining on those pages that use `#REDIRECT`.

## Solution

I 'd like both of you to test the following change to `SMWSQLStore3Writers::doFlatDataUpdate` in any possible way and confirm that redirects via `#REDIRECT` or `Special:MovePage` create a consistence behaviour indifferent of the method used.

You have to add a temporary hack (lines marked with `+`) to  `SMWSQLStore3Writers::doFlatDataUpdate` [0] before this can be turned into a real patch (after it is confirmed that it works).

```php
+++ b/includes/storage/SQLStore/SMW_SQLStore3_Writers.php
@@ -155,19 +155,50 @@ class SMWSQLStore3Writers {
 
 		if ( count( $redirects ) > 0 ) {
 			$redirect = end( $redirects ); // at most one redirect per page
+
+			// Find out whether the redirect is known or not (e.g use of #REDIRECT
+			// for a non-existing page)
+			$tid = $this->store->getObjectIds()->getSMWPageID(
+				$redirect->getDBkey(),
+				$redirect->getNamespace(),
+				'',
+				'',
+				false
+			);
+
 			$this->updateRedirects(
 				$subject->getDBkey(),
 				$subject->getNamespace(),
 				$redirect->getDBkey(),
 				$redirect->getNameSpace()
 			);
+
 			// Stop here:
 			// * no support for annotations on redirect pages
 			// * updateRedirects takes care of deleting any previous data
-
 			$this->store->getConnection()->commitTransaction( __METHOD__ );
 
-			return;
+			// Bail-out if the redirect is known otherwise copy the reference and
+			// continue to ensure an appropriate clean-up
+			if ( $tid != 0 ) {
+				return;
+			} else {
+				$this->store->getConnection()->update(
+					SMWSql3SmwIds::tableName,
+					array(
+						'smw_title' => $redirect->getDBkey(),
+						'smw_namespace' => $redirect->getNamespace(),
+						'smw_iw' => ''
+					),
+					array(
+						'smw_title' => $subject->getDBkey(),
+						'smw_namespace' => $subject->getNamespace(),
+						'smw_iw' => ''
+					),
+					__METHOD__
+				);
+			}
+
 		} else {
```

## Tests
- Use  `#REDIRECT`/`Special:MovePage` for known/unknown targets (exists/doesn't exist)
- Content with/without use of template annotations
- Run `rebuildData` update jobs
- Check query output with focus on the correct target 
- If data is moved via `Special:MovePage` annotations should be shown on the target page as well as the original source (because the annotations/content is copied)
- If data is moved using `#REDIRECT` data should be cleared (because you have the remove all content from the source and leave only a single internal `_REDI` reference)

I know this is a bit a work but I'd like to avoid a situation where we introduce a regression just under the impression to have solved a behavioral inconsistency.

[0] https://github.com/SemanticMediaWiki/SemanticMediaWiki/blob/master/includes/storage/SQLStore/SMW_SQLStore3_Writers.php#L156

**Comment by mwjames** [comment]
Any update?

**Comment by kghbln** [comment]
### Page with &action=move to redirect to unknown location
* Created page ABC/1 with {{UseTemplateForAnnotation}} content
* Moved page ABC/1 to an unknown ABC/2 (left a redirect behind)

I ended up with no outgoing (not even special properties) and one incoming annotation (has pagename ABC/2) on ABC/1 and no annotations (not even special properties) on ABC/2 (only telling about the redirect from ABC/1)

After the data rebuilding: Absolutely nothing on page ABC/1 and all properties including special properties on page ABC/2 including a duplicated incoming property values (has pagename) as well as the info about the redirect from ABC/1.

### Page with &action=move to redirect to known location
* Created page ABC/3 with {{UseTemplateForAnnotation}} content
* Moved page ABC/3 to known ABC/4 (deleted page ABC/4 to allow moving ABC/3 to ABC/4)

I ended up with no outgoing (not even special properties) and one incoming annotation (has pagename ABC/4) on ABC/4 and no annotations (not even special properties) on ABC/4 (only telling about the redirect from ABC/3) = same as first case

After the data rebuilding: same as first case

### Page with #REDIRECT edit to redirect to unknown location
* Created page DEF/1 with {{UseTemplateForAnnotation}} content
* Edited page DEF/1 to unknown #REDIRECT [[DEF/2]]

I ended up with no outgoing (not even special properties) and NO incoming annotation on DEF/1 and only special property annotations on DEF/2  and one incoming property (has pagename DEF/2) and no info about a redirect from DEF/1.

After the data rebuilding: Absolutely nothing on page DEF/1 and DEF/2 (only an info about the redirect from DEF/1)

### Page with #REDIRECT edit to redirect to known location
* Created page DEF/3 with {{UseTemplateForAnnotation}} content
* Edited page DEF/3 to known #REDIRECT [[DEF/4]]

I ended up with all original annotations (including special properties) and one incoming propery (has pagename DEF/3). Besides, according to the special properties the page thinks that it is still untouched since the first edit. On page DEF/4 is still have the original special properties as well as the original regular outgoing properties. No incoming property here.

After the data rebuilding: As if no data rebuilding had happend. Same behaviour as before the rebuilding for these two pages.

I am on MW 1.23.8 and I am not using equality support. Semantic extensions are on master except for the changes to the above mentioned file. Hope this helps a bit.

**Comment by mwjames** [comment]
Expected to be solved by c828143
